### DISCUSSION TITLE
Annual and Monthly archive.

### DISCUSSION THREAD

**Discussion by shobhithn7** [statamic/cms#5487] [Q&A]
Does anybody have any good code snippets for producing a dynamic list of archives by year > month ?

**Comment by pepijndevos** [comment]
This sad and lonely post turned up at the top of the search result when I tried to make monthly archives with Statamic, and after a day of suffering I thought I'd relieve the next web wanderer.

First you need to set up routes that redirect the archive urls to a template

routes/web.php
```
Route::statamic('subjects/{subject}/{year}/{month}', 'subjects.archive');
Route::statamic('subjects/{subject}/{year}', 'subjects.archive');
```

and then in the template you can do

```
    {{ collection:activities 
        taxonomy:subjects="{{ slug }}"
        since="{{ since_date }}" 
        until="{{ until_date }}"
        sort="date:desc"
    }}
```

but getting the dates correct was a bit of a mindfuck since they appear to be exclusive ranges, and there doesn't seem to be a good way to turn string variables into date objects to manipulate. But the good news is until/since use PHP's `strtotime` which supports all sorts of exotic combinations so in the end I used something like

```
    {{ if year && month }}
        {{ 
            since_date = year + '-' + month + ' last day of previous month';
            until_date = year + '-' + month + ' first day of next month';
        }}
    {{ elseif year }}
        {{ 
            since_date = year + ' last day of previous year';
            until_date = year + ' first day of next year';
        }}
    {{ else }}
        {{ 
            since_date = '';
            until_date = '';
        }}
    {{ /if }}
```

that is, it gives it a year or year-month and then combines that with the relative date arithmetic in string form. You gotta love PHP for supporting outlandish stuff like this in the standard library. So you'd get something like "2025-01 last day of previous month" which would result in the date 2024-12-31 so that you include the first day of the year in your archive range.

Maybe I'm missing something obvious like an "inclusive=true" flag somewhere but that's what I landed on. Hope it helps someone.

For generating the actual list you can just use group_by like so

```
    {{ collection:activities taxonomy:subjects="{{ slug }}" as="activities" }}
        {{ activities group_by="date|F Y" }}
            {{ groups }}
```

I haven't managed to have a nested loop of year->month but I haven't tried very hard.

**Comment by ebeauchamps** [comment]
[Dates](https://statamic.dev/modifiers/group_by#dates)
You may group entries by a date field.

For example, you might want to output articles and group them by month.

Here we'll group them by the date field using the F Y [PHP date format](https://www.php.net/manual/en/datetime.format.php) which
would output as "Month Year".
```
{{ collection:articles as="entries" }}
    {{ entries group_by="date|F Y" }}
        {{ groups }}
            <h1>{{ group }}</h1>
            {{ items }}
                {{ title }}
            {{ /items }}
        {{ /groups }}
    {{ /entries }}
{{ /collection:articles }}
```

HOT TIP!
The date field in this example is named date, but you can use any date field.
```
{{ entries group_by="custom_date_field|F Y" }}
```

If you need the key to differ from how it's displayed (perhaps you want to use an additional modifier after), you can pass another date format as the
third argument. (Argument 2 creates the key, argument 3 creates the {{ group }} text).
```
{{ entries group_by="date|Y-m|F Y" }}
```
